azure-iot-device/tests/common/pipeline/pipeline_ops_test.py

```diff
@@ -0,0 +1,461 @@
+
+    :param test_module: A reference to the test module to add tests to
+    :param op_class_under_test: A reference to the specific Operation class under test
+    :param op_test_config_class: A class providing fixtures specific to the Operation class
+        under test. This class must define the following fixtures:
+            - "cls_type" (which returns a reference to the Operation class under test)
+            - "init_kwargs" (which returns a dictionary of kwargs and associated values used to
+                instantiate the class)
+    :param extended_op_instantiation_test_class: A class defining instantiation tests that are
+        specific to the Operation class under test, and not shared with all Operations.
+        Note that you may override shared instantiation tests defined in this function within
+        the provided test class (e.g. test_needs_connection)
+    """
+
+    # Extend the provided test config class
+    class OperationTestConfigClass(op_test_config_class):
+        @pytest.fixture
+        def op(self, cls_type, init_kwargs, mocker):
+            op = cls_type(**init_kwargs)
+            mocker.spy(op, "complete")
+            return op
+
+    @pytest.mark.describe("{} - Instantiation".format(op_class_under_test.__name__))
+    class OperationBaseInstantiationTests(OperationTestConfigClass):
+        @pytest.mark.it("Initializes 'name' attribute as the classname")
+        def test_name(self, cls_type, init_kwargs):
+            op = cls_type(**init_kwargs)
+            assert op.name == op.__class__.__name__
+
+        @pytest.mark.it("Initializes 'completed' attribute as False")
+        def test_completed(self, cls_type, init_kwargs):
+            op = cls_type(**init_kwargs)
+            assert op.completed is False
+
+        # NOTE: this test should be overridden for operations that set this value to True
+        @pytest.mark.it("Initializes 'needs_connection' attribute as False")
+        def test_needs_connection(self, cls_type, init_kwargs):
+            op = cls_type(**init_kwargs)
+            assert op.needs_connection is False
+
+        @pytest.mark.it("Initializes 'callbacks' list attribute with the provided callback")
+        def test_callback_added_to_list(self, cls_type, init_kwargs):
+            op = cls_type(**init_kwargs)
+            assert len(op.callbacks) == 1
+            assert op.callbacks[0] is init_kwargs["callback"]
+
+    # If an extended operation instantiation test class is provided, use those tests as well.
+    # By using the extended_op_instantation_test_class as the first parent class, this ensures that
+    # tests from OperationBaseInstantiationTests (e.g. test_needs_connection) can be overwritten by
+    # tests provided in extended_op_instantiation_test_class.
+    if extended_op_instantiation_test_class:
+
+        class OperationInstantiationTests(
+            extended_op_instantiation_test_class, OperationBaseInstantiationTests
```

Comment:
>extended_op_instantiation_test_class [](start = 12, length = 36)

Language question: if you want a method implementation in extended_op_initialization_test_class to "replace" the implementation in OperationBaseInstantiationTests, wouldn't you reverse the order of these classes?  Doesn't the second base class get "applied" after the first?